polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -430,353 +452,207 @@ impl Scope {
-	///
-	/// Can be populated recursively (i.e. `populate` will pick up candidates that build on other
-	/// candidates).
+impl FragmentChain {
+	/// Create a new [`FragmentChain`] with given scope and populated from the storage.
 	pub fn populate(scope: Scope, storage: &CandidateStorage) -> Self {
 		gum::trace!(
 			target: LOG_TARGET,
 			relay_parent = ?scope.relay_parent.hash,
 			relay_parent_num = scope.relay_parent.number,
 			para_id = ?scope.para,
 			ancestors = scope.ancestors.len(),
-			"Instantiating Fragment Tree",
+			"Instantiating Fragment Chain",
 		);
 
-		let mut tree = FragmentTree { scope, nodes: Vec::new(), candidates: HashMap::new() };
+		let mut fragment_chain = Self {
+			scope,
+			chain: Vec::new(),
+			candidates: HashSet::new(),
+			by_parent_head: HashMap::new(),
+			by_output_head: HashMap::new(),
+		};
 
-		tree.populate_from_bases(storage, vec![NodePointer::Root]);
+		fragment_chain.populate_chain(storage);
 
-		tree
+		fragment_chain
 	}
 
-	/// Get the scope of the Fragment Tree.
+	/// Get the scope of the Fragment Chain.
 	pub fn scope(&self) -> &Scope {
 		&self.scope
 	}
 
-	// Inserts a node and updates child references in a non-root parent.
-	fn insert_node(&mut self, node: FragmentNode) {
-		let pointer = NodePointer::Storage(self.nodes.len());
-		let parent_pointer = node.parent;
-		let candidate_hash = node.candidate_hash;
-
-		let max_depth = self.scope.max_depth;
-
-		self.candidates
-			.entry(candidate_hash)
-			.or_insert_with(|| bitvec![u16, Msb0; 0; max_depth + 1])
-			.set(node.depth, true);
-
-		match parent_pointer {
-			NodePointer::Storage(ptr) => {
-				self.nodes.push(node);
-				self.nodes[ptr].children.push((pointer, candidate_hash))
-			},
-			NodePointer::Root => {
-				// Maintain the invariant of node storage beginning with depth-0.
-				if self.nodes.last().map_or(true, |last| last.parent == NodePointer::Root) {
-					self.nodes.push(node);
-				} else {
-					let pos =
-						self.nodes.iter().take_while(|n| n.parent == NodePointer::Root).count();
-					self.nodes.insert(pos, node);
-				}
-			},
-		}
-	}
-
-	fn node_has_candidate_child(
-		&self,
-		pointer: NodePointer,
-		candidate_hash: &CandidateHash,
-	) -> bool {
-		self.node_candidate_child(pointer, candidate_hash).is_some()
-	}
-
-	fn node_candidate_child(
-		&self,
-		pointer: NodePointer,
-		candidate_hash: &CandidateHash,
-	) -> Option<NodePointer> {
-		match pointer {
-			NodePointer::Root => self
-				.nodes
-				.iter()
-				.take_while(|n| n.parent == NodePointer::Root)
-				.enumerate()
-				.find(|(_, n)| &n.candidate_hash == candidate_hash)
-				.map(|(i, _)| NodePointer::Storage(i)),
-			NodePointer::Storage(ptr) =>
-				self.nodes.get(ptr).and_then(|n| n.candidate_child(candidate_hash)),
-		}
-	}
-
 	/// Returns an O(n) iterator over the hashes of candidates contained in the
 	/// tree.
 	pub(crate) fn candidates(&self) -> impl Iterator<Item = CandidateHash> + '_ {
```

Comment:
Does the caller always have to clone the hashes ? If not, and hope for we should return iterator over references.

Reply:
replaced this with the `to_vec` method on the FragmentChain